File: sdk/typescript/src/namespaces/disaster-recovery.ts
/**
 * Disaster Recovery Namespace API
 *
 * Provides a namespaced interface for disaster recovery operations.
 * This wraps the flat client methods for DR drills, status, and validation.
 */

/**
 * DR status response.
 */
export interface DRStatus {
  ready: boolean;
  last_backup_at: string | null;
  last_drill_at: string | null;
  rpo_met: boolean;
  rto_met: boolean;
  issues: DRIssue[];
  overall_health: 'healthy' | 'degraded' | 'critical';
}

/**
 * DR issue detail.
 */
export interface DRIssue {
  severity: 'low' | 'medium' | 'high' | 'critical';
  component: string;
  message: string;
  remediation: string;
}

/**
 * Recovery objectives.
 */
export interface DRObjectives {
  rpo_minutes: number;
  rto_minutes: number;
  current_rpo_minutes: number;
  current_rto_minutes: number;
  rpo_compliant: boolean;
  rto_compliant: boolean;
  last_measured_at: string;
}

/**
 * DR drill request.
 */
export interface DRDrillRequest {
  type: 'tabletop' | 'simulation' | 'full';
  components?: string[];
  notify_team?: boolean;
  dry_run?: boolean;
}

/**
 * DR drill result.
 */
export interface DRDrillResult {
  drill_id: string;
  type: string;
  started_at: string;
  completed_at: string;
  duration_seconds: number;
  success: boolean;
  recovery_time_seconds: number;
  data_loss_seconds: number;
  steps: DRDrillStep[];
  recommendations: string[];
}

/**
 * DR drill step detail.
 */
export interface DRDrillStep {
  name: string;
  status: 'passed' | 'failed' | 'skipped';
  duration_seconds: number;
  message: string;
}

/**
 * DR validation request.
 */
export interface DRValidateRequest {
  check_backups?: boolean;
  check_replication?: boolean;
  check_failover?: boolean;
  check_dns?: boolean;
}

/**
 * DR validation result.
 */
export interface DRValidationResult {
  valid: boolean;
  checks: DRValidationCheck[];
  overall_score: number;
  last_validated_at: string;
}

/**
 * Individual validation check.
 */
export interface DRValidationCheck {
  name: string;
  passed: boolean;
  message: string;
  details: Record<string, unknown>;
}

/**
 * Interface for the internal client methods used by DisasterRecoveryAPI.
 */
interface DRClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: unknown }
  ): Promise<T>;
}

/**
 * Disaster Recovery API namespace.
 *
 * Provides methods for enterprise disaster recovery operations:
 * - Checking DR readiness status
 * - Running DR drills (tabletop, simulation, full)
 * - Getting and monitoring RPO/RTO objectives
 * - Validating DR configuration
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // Check DR status
 * const status = await client.dr.getStatus();
 * console.log(`DR ready: ${status.ready}, Health: ${status.overall_health}`);
 *
 * // Run a tabletop drill
 * const drill = await client.dr.runDrill({
 *   type: 'tabletop',
 *   notify_team: true,
 * });
 *
 * // Check recovery objectives
 * const objectives = await client.dr.getObjectives();
 * console.log(`RPO: ${objectives.rpo_minutes}min, RTO: ${objectives.rto_minutes}min`);
 * ```
 */
export class DisasterRecoveryAPI {
  constructor(private client: DRClientInterface) {}

  /**
   * Get current DR readiness status.
   */
  async getStatus(): Promise<DRStatus> {
    return this.client.request<DRStatus>('GET', '/api/v2/dr/status');
  }

  /**
   * Run a DR drill.
   *
   * @param request - Drill configuration
   * @returns Drill results including recovery time and data loss metrics
   */
  async runDrill(request: DRDrillRequest): Promise<DRDrillResult> {
    return this.client.request<DRDrillResult>('POST', '/api/v2/dr/drill', {
      json: request,
    });
  }

  /**
   * Get current RPO/RTO objectives and compliance status.
   */
  async getObjectives(): Promise<DRObjectives> {
    return this.client.request<DRObjectives>('GET', '/api/v2/dr/objectives');
  }

  /**
   * Validate DR configuration.
   *
   * @param request - Optional validation scope
   * @returns Validation results with detailed check status
   */
  async validate(request?: DRValidateRequest): Promise<DRValidationResult> {
    return this.client.request<DRValidationResult>('POST', '/api/v2/dr/validate', {
      json: request || {},
    });
  }

  /**
   * Check if DR is ready for production.
   * Convenience method that checks status and validates configuration.
   */
  async isReady(): Promise<boolean> {
    const [status, validation] = await Promise.all([
      this.getStatus(),
      this.validate(),
    ]);
    return status.ready && validation.valid;
  }

  /**
   * Get a summary of DR health.
   */
  async getHealthSummary(): Promise<{
    ready: boolean;
    health: string;
    rpo_compliant: boolean;
    rto_compliant: boolean;
    issues_count: number;
  }> {
    const [status, objectives] = await Promise.all([
      this.getStatus(),
      this.getObjectives(),
    ]);
    return {
      ready: status.ready,
      health: status.overall_health,
      rpo_compliant: objectives.rpo_compliant,
      rto_compliant: objectives.rto_compliant,
      issues_count: status.issues.length,
    };
  }
}

export default DisasterRecoveryAPI;

File: sdk/python/aragora_sdk/namespaces/disaster_recovery.py
"""
Disaster Recovery Namespace API

Provides methods for disaster recovery operations:
- Checking DR readiness status
- Running DR drills (tabletop, simulation, full)
- Getting and monitoring RPO/RTO objectives
- Validating DR configuration
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class DisasterRecoveryAPI:
    """
    Synchronous Disaster Recovery API.

    Provides methods for enterprise disaster recovery operations:
    - Backup management (create, list, get, delete)
    - Backup verification and restore testing
    - Cleanup and statistics

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> backups = client.dr.list_backups()
        >>> stats = client.dr.get_stats()
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # ===========================================================================
    # Backups (v2)
    # ===========================================================================

    def list_backups(self) -> dict[str, Any]:
        """List all backups."""
        return self._client.request("GET", "/api/v2/backups")

    def create_backup(self, **kwargs: Any) -> dict[str, Any]:
        """Create a new backup."""
        return self._client.request("POST", "/api/v2/backups", json=kwargs)

    def get_backup(self, backup_id: str) -> dict[str, Any]:
        """Get a backup by ID."""
        return self._client.request("GET", f"/api/v2/backups/{backup_id}")

    def delete_backup(self, backup_id: str) -> dict[str, Any]:
        """Delete a backup."""
        return self._client.request("DELETE", f"/api/v2/backups/{backup_id}")

    # ===========================================================================
    # Backup Operations (v1)
    # ===========================================================================

    def get_stats(self) -> dict[str, Any]:
        """Get backup statistics."""
        return self._client.request("GET", "/api/v1/backups/stats")

    def cleanup(self, **kwargs: Any) -> dict[str, Any]:
        """Run backup cleanup."""
        return self._client.request("POST", "/api/v1/backups/cleanup", json=kwargs)

    def verify(self, backup_id: str) -> dict[str, Any]:
        """Verify a backup."""
        return self._client.request("POST", f"/api/v1/backups/{backup_id}/verify")

    def verify_comprehensive(self, backup_id: str) -> dict[str, Any]:
        """Run comprehensive backup verification."""
        return self._client.request("POST", f"/api/v1/backups/{backup_id}/verify-comprehensive")

    def restore_test(self, backup_id: str) -> dict[str, Any]:
        """Run a restore test on a backup."""
        return self._client.request("POST", f"/api/v1/backups/{backup_id}/restore-test")

    # ===========================================================================
    # Convenience Methods
    # ===========================================================================

    def is_ready(self) -> bool:
        """Check if DR is ready for production."""
        status = self.get_status()
        validation = self.validate()
        return bool(status.get("ready")) and bool(validation.get("valid"))

    def get_health_summary(self) -> dict[str, Any]:
        """Get a summary of DR health."""
        status = self.get_status()
        objectives = self.get_objectives()
        return {
            "ready": status.get("ready", False),
            "health": status.get("overall_health", "unknown"),
            "rpo_compliant": objectives.get("rpo_compliant", False),
            "rto_compliant": objectives.get("rto_compliant", False),
            "issues_count": len(status.get("issues", [])),
        }


class AsyncDisasterRecoveryAPI:
    """
    Asynchronous Disaster Recovery API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     backups = await client.dr.list_backups()
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # ===========================================================================
    # Backups (v2)
    # ===========================================================================

    async def list_backups(self) -> dict[str, Any]:
        """List all backups."""
        return await self._client.request("GET", "/api/v2/backups")

    async def create_backup(self, **kwargs: Any) -> dict[str, Any]:
        """Create a new backup."""
        return await self._client.request("POST", "/api/v2/backups", json=kwargs)

    async def get_backup(self, backup_id: str) -> dict[str, Any]:
        """Get a backup by ID."""
        return await self._client.request("GET", f"/api/v2/backups/{backup_id}")

    async def delete_backup(self, backup_id: str) -> dict[str, Any]:
        """Delete a backup."""
        return await self._client.request("DELETE", f"/api/v2/backups/{backup_id}")

    # ===========================================================================
    # Backup Operations (v1)
    # ===========================================================================

    async def get_stats(self) -> dict[str, Any]:
        """Get backup statistics."""
        return await self._client.request("GET", "/api/v1/backups/stats")

    async def cleanup(self, **kwargs: Any) -> dict[str, Any]:
        """Run backup cleanup."""
        return await self._client.request("POST", "/api/v1/backups/cleanup", json=kwargs)

    async def verify(self, backup_id: str) -> dict[str, Any]:
        """Verify a backup."""
        return await self._client.request("POST", f"/api/v1/backups/{backup_id}/verify")

    async def verify_comprehensive(self, backup_id: str) -> dict[str, Any]:
        """Run comprehensive backup verification."""
        return await self._client.request("POST", f"/api/v1/backups/{backup_id}/verify-comprehensive")

    async def restore_test(self, backup_id: str) -> dict[str, Any]:
        """Run a restore test on a backup."""
        return await self._client.request("POST", f"/api/v1/backups/{backup_id}/restore-test")

    # ===========================================================================
    # Convenience Methods
    # ===========================================================================

    async def is_ready(self) -> bool:
        """Check if DR is ready for production."""
        status = await self.get_status()
        validation = await self.validate()
        return bool(status.get("ready")) and bool(validation.get("valid"))

    async def get_health_summary(self) -> dict[str, Any]:
        """Get a summary of DR health."""
        status = await self.get_status()
        objectives = await self.get_objectives()
        return {
            "ready": status.get("ready", False),
            "health": status.get("overall_health", "unknown"),
            "rpo_compliant": objectives.get("rpo_compliant", False),
            "rto_compliant": objectives.get("rto_compliant", False),
            "issues_count": len(status.get("issues", [])),
        }
